src/adapter/adapter.ts

```diff
@@ -47,15 +48,28 @@ abstract class Adapter extends events.EventEmitter {
             }
         } else {
             try {
-                // Path can be a symlink, resolve it.
-                serialPortOptions.path = RealpathSync(serialPortOptions.path);
-
-                // Determine adapter to use
-                for (const candidate of adapters) {
-                    if (await candidate.isValidPath(serialPortOptions.path)) {
-                        debug(`Path '${serialPortOptions.path}' is valid for '${candidate.name}'`);
-                        adapter = candidate;
-                        break;
+                if (SocketPortUtils.isTcp(serialPortOptions.path)) {
+                    // check socket port
+                    if (SocketPortUtils.isValidTcpPath(serialPortOptions.path)) {
+                        debug(`Path '${serialPortOptions.path}' is valid`);
+                    }
+                    else {
+                        throw new Error(`Invalid path: '${serialPortOptions.path}'`);
+                    }
+                }
+                else {
+                    // check serial port
+
+                    // Path can be a symlink, resolve it.
+                    serialPortOptions.path = RealpathSync(serialPortOptions.path);
+
+                    // Determine adapter to use
+                    for (const candidate of adapters) {
+                        if (await candidate.isValidPath(serialPortOptions.path)) {
```

Comment:
I don't feel there should be separate branches here, `isValidPath` should also provide the `isValidTcpPath` functionallity. Why has this been separated?

Reply:
I first used that but 'RealpathSync' does not work with 'tcp://' entries. You're right this could be also checked at 'isValidPath' but then RealpathSync should be skipped for 'tcp://' entries.

Reply:
Yes please, realPathSync should be moved to the adapter itself I think

Reply:
I reverted the changes on adapter.ts. However to prevent potential side effects with symbolic links on serial paths i modified 'realpathsync' to ignore 'tcp:' path syncing. I also improved tcp path validation in znp.ts with proper exception throwing. Moving 'realpathsync' from adapter.ts to znp.ts would be much more complicated and prone to side effects. What do you think?

Reply:
Why would this be much more complicated? Now `realpathsync` has unexpected behaviour when passing `tcp:` paths.

Reply:
You're right. I reverted realpathsync and moved the calling method from adapter.ts to znp. It's in the constructor of znp now and also in the static isValidPath method of znp.

Reply:
Good, how can I make a test setup?

Reply:
You can test this locally with the ser2net package:

(see https://linux.die.net/man/8/ser2net)

1) Install ser2net package

`sudo apt-get install ser2net`

On my Ubuntu 18.04 LTS only version 2.10 is available but works fine.

2) Edit configuration to bind serial port to network port

`sudo vi /etc/ser2net.conf`

Add following configuration (could also be /dev/ttyACM0 depending on your machine)

`20108:raw:0:/dev/ttyACM1:115200 8DATABITS NONE 1STOPBIT`


3) Edit zigbe2mqtt configuration to use network port

```
vi ./data/configuration.yaml
port: 'tcp://127.0.0.1:20108'
```


Note: The string 'localhost' is currently invalid so i'm using '127.0.0.1'. I'll fix this.


4) Start zigbee2mqtt

`npm start`

Reply:
Ok thanks, so last thing to do here is add the tests (so it passes CICD checks).

Reply:
Alright, i'll try to add the missing tests later today :-)